File: macosHelper/Sources/PosatoMacOSServiceCore/BrowserDomainConfigureProtocol.swift
import Foundation

public enum BrowserDomainConfigureLimits {
  public static let maximumDomainCount = 1_024
  public static let minimumDomainLength = 3
  public static let maximumDomainLength = 253
  public static let minimumLabelLength = 1
  public static let maximumLabelLength = 63
}

public struct BrowserDomainConfigurePayload: Equatable, Sendable {
  public let domains: [String]
  public let sessionEndEpochMilliseconds: UInt64?

  public init(domains: [String], sessionEndEpochMilliseconds: UInt64?) throws {
    guard domains.count >= 1,
      domains.count <= BrowserDomainConfigureLimits.maximumDomainCount,
      Set(domains).count == domains.count,
      domains.allSatisfy(Self.isCanonicalExactDomain)
    else {
      throw WireProtocolFailure.invalidFrame
    }
    self.domains = domains
    self.sessionEndEpochMilliseconds = sessionEndEpochMilliseconds
  }

  public func encode() throws -> Data {
    var data = Data()
    data.appendBigEndian(UInt16(domains.count))
    for domain in domains {
      let encoded = Data(domain.utf8)
      guard encoded.count <= UInt8.max else {
        throw WireProtocolFailure.invalidFrame
      }
      data.append(UInt8(encoded.count))
      data.append(encoded)
    }
    if let sessionEndEpochMilliseconds {
      data.append(1)
      data.appendBigEndian(sessionEndEpochMilliseconds)
    } else {
      data.append(0)
    }
    guard data.count <= WireLimits.maximumFrameBytes - 68 else {
      throw WireProtocolFailure.oversizedFrame
    }
    return data
  }

  public static func decode(_ data: Data) throws -> BrowserDomainConfigurePayload {
    var cursor = WireDataCursor(data: data)
    let count = Int(try cursor.readUInt16())
    guard count >= 1, count <= BrowserDomainConfigureLimits.maximumDomainCount else {
      throw WireProtocolFailure.invalidFrame
    }
    var domains: [String] = []
    domains.reserveCapacity(count)
    for _ in 0..<count {
      let domainData = try cursor.readBoundedData(
        maximumBytes: BrowserDomainConfigureLimits.maximumDomainLength
      )
      guard let domain = String(data: domainData, encoding: .utf8) else {
        throw WireProtocolFailure.invalidFrame
      }
      domains.append(domain)
    }
    let hasEndTime = try cursor.readUInt8()
    let sessionEndEpochMilliseconds: UInt64?
    switch hasEndTime {
    case 0:
      sessionEndEpochMilliseconds = nil
    case 1:
      sessionEndEpochMilliseconds = try cursor.readUInt64()
    default:
      throw WireProtocolFailure.invalidFrame
    }
    guard cursor.remainingBytes == 0 else {
      throw WireProtocolFailure.invalidFrame
    }
    return try BrowserDomainConfigurePayload(
      domains: domains,
      sessionEndEpochMilliseconds: sessionEndEpochMilliseconds
    )
  }

  private static func isCanonicalExactDomain(_ domain: String) -> Bool {
    let length = domain.utf8.count
    guard length >= BrowserDomainConfigureLimits.minimumDomainLength,
      length <= BrowserDomainConfigureLimits.maximumDomainLength,
      domain.allSatisfy({ character in
        character.isASCII
          && ((character.isLetter && character.isLowercase) || character.isNumber
            || character == "-" || character == ".")
      })
    else {
      return false
    }
    let labels = domain.split(separator: ".", omittingEmptySubsequences: false)
    guard labels.count >= 2, let last = labels.last, !last.allSatisfy(\.isNumber) else {
      return false
    }
    return labels.allSatisfy(isCanonicalDomainLabel)
  }

  private static func isCanonicalDomainLabel(_ label: Substring) -> Bool {
    let length = label.utf8.count
    guard length >= BrowserDomainConfigureLimits.minimumLabelLength,
      length <= BrowserDomainConfigureLimits.maximumLabelLength,
      let first = label.first, first.isLetter || first.isNumber,
      let last = label.last, last.isLetter || last.isNumber
    else {
      return false
    }
    return label.allSatisfy { character in
      character.isLetter || character.isNumber || character == "-"
    }
  }
}

extension BrowserDomainConfigurePayload: CustomStringConvertible, CustomDebugStringConvertible {
  public var description: String {
    return "BrowserDomainConfigurePayload(redacted)"
  }

  public var debugDescription: String {
    return description
  }
}

public struct BrowserDomainConfigureResponse: Equatable, Sendable {
  public let outcome: WireResponsePayload
  public let port: UInt16

  public init(outcome: WireResponsePayload, port: UInt16) throws {
    let successRequiresPort = outcome.outcome == .success
    guard successRequiresPort ? port > 0 : port == 0 else {
      throw WireProtocolFailure.invalidFrame
    }
    self.outcome = outcome
    self.port = port
  }

  public func encode() -> Data {
    var data = outcome.encode()
    data.appendBigEndian(port)
    return data
  }

  public static func decode(_ data: Data) throws -> BrowserDomainConfigureResponse {
    guard data.count == 7 else {
      throw WireProtocolFailure.invalidFrame
    }
    let outcome = try WireResponsePayload.decode(Data(data.prefix(5)))
    let port = data.suffix(2).reduce(UInt16(0)) { value, byte in
      (value << 8) | UInt16(byte)
    }
    return try BrowserDomainConfigureResponse(outcome: outcome, port: port)
  }
}

extension BrowserDomainConfigureResponse: CustomStringConvertible, CustomDebugStringConvertible {
  public var description: String {
    return "BrowserDomainConfigureResponse(redacted)"
  }

  public var debugDescription: String {
    return description
  }
}

File: desktopApp/src/main/kotlin/app/posato/desktop/macos/BrowserDomainConfigureProtocol.kt
package app.posato.desktop.macos

import java.nio.ByteBuffer
import java.nio.ByteOrder

internal object BrowserDomainConfigureLimits {
    const val MAXIMUM_DOMAIN_COUNT: Int = 1_024
    const val MINIMUM_DOMAIN_LENGTH: Int = 3
    const val MAXIMUM_DOMAIN_LENGTH: Int = 253
    const val MINIMUM_LABEL_LENGTH: Int = 1
    const val MAXIMUM_LABEL_LENGTH: Int = 63
    const val MAXIMUM_ASCII_CODE: Int = 128
    const val BYTE_MASK: Int = 0xFF
    const val PORT_BYTE_SHIFT: Int = 8
    const val RESPONSE_SIZE_BYTES: Int = 7
    const val RESPONSE_PORT_HIGH_INDEX: Int = 5
    const val RESPONSE_PORT_LOW_INDEX: Int = 6
    const val RESULT_SIZE_BYTES: Int = 5
    const val NO_END_TIME_FLAG: Int = 0
    const val HAS_END_TIME_FLAG: Int = 1
}

internal data class BrowserDomainConfigurePayload(
    val domains: List<String>,
    val sessionEndEpochMilliseconds: Long?,
) {
    init {
        require(domains.size in 1..BrowserDomainConfigureLimits.MAXIMUM_DOMAIN_COUNT)
        require(domains.toSet().size == domains.size)
        require(domains.all(::isCanonicalExactDomain))
        sessionEndEpochMilliseconds?.let { end ->
            require(end >= 0)
        }
    }

    fun encode(): ByteArray {
        val encodedDomains = domains.map { domain -> domain.encodeToByteArray() }
        val endTimeBytes = if (sessionEndEpochMilliseconds == null) 1 else 9
        val size = 2 + encodedDomains.sumOf { encoded -> 1 + encoded.size } + endTimeBytes
        val buffer = ByteBuffer.allocate(size).order(ByteOrder.BIG_ENDIAN)
        buffer.putShort(encodedDomains.size.toShort())
        encodedDomains.forEach { encoded ->
            require(encoded.size in 1..UByte.MAX_VALUE.toInt())
            buffer.put(encoded.size.toByte())
            buffer.put(encoded)
        }
        if (sessionEndEpochMilliseconds == null) {
            buffer.put(0)
        } else {
            buffer.put(1)
            buffer.putLong(sessionEndEpochMilliseconds)
        }
        return buffer.array()
    }

    override fun toString(): String {
        return "BrowserDomainConfigurePayload(redacted)"
    }

    companion object {
        fun decode(payload: ByteArray): BrowserDomainConfigurePayload {
            val buffer = ByteBuffer.wrap(payload).order(ByteOrder.BIG_ENDIAN)
            require(buffer.remaining() >= 2)
            val count = buffer.short.toInt() and 0xFFFF
            require(count in 1..BrowserDomainConfigureLimits.MAXIMUM_DOMAIN_COUNT)
            val domains = buildList {
                repeat(count) {
                    require(buffer.remaining() >= 1)
                    val length = buffer.get().toInt() and BrowserDomainConfigureLimits.BYTE_MASK
                    require(length in 1..BrowserDomainConfigureLimits.MAXIMUM_DOMAIN_LENGTH)
                    require(buffer.remaining() >= length)
                    val encoded = ByteArray(length)
                    buffer.get(encoded)
                    add(encoded.decodeToString())
                }
            }
            require(buffer.remaining() >= 1)
            val hasEndTime = buffer.get().toInt() and BrowserDomainConfigureLimits.BYTE_MASK
            val sessionEndEpochMilliseconds = readSessionEndTime(buffer, hasEndTime)
            require(!buffer.hasRemaining())
            return BrowserDomainConfigurePayload(domains, sessionEndEpochMilliseconds)
        }

        private fun readSessionEndTime(
            buffer: ByteBuffer,
            hasEndTime: Int,
        ): Long? {
            if (hasEndTime == BrowserDomainConfigureLimits.NO_END_TIME_FLAG) {
                return null
            }
            if (hasEndTime == BrowserDomainConfigureLimits.HAS_END_TIME_FLAG) {
                require(buffer.remaining() >= Long.SIZE_BYTES)
                return buffer.long
            }
            error("Invalid configure end-time flag")
        }
    }
}

internal data class BrowserDomainConfigureResponse(
    val result: HelperResult,
    val port: UShort,
) {
    init {
        if (result.outcome == HelperResult.Outcome.Success) {
            require(port > 0u)
        } else {
            require(port == 0.toUShort())
        }
    }

    override fun toString(): String {
        return "BrowserDomainConfigureResponse(redacted)"
    }

    companion object {
        fun decode(payload: ByteArray): BrowserDomainConfigureResponse {
            require(payload.size == BrowserDomainConfigureLimits.RESPONSE_SIZE_BYTES)
            val high = payload[BrowserDomainConfigureLimits.RESPONSE_PORT_HIGH_INDEX].toInt() and
                BrowserDomainConfigureLimits.BYTE_MASK
            val low = payload[BrowserDomainConfigureLimits.RESPONSE_PORT_LOW_INDEX].toInt() and
                BrowserDomainConfigureLimits.BYTE_MASK
            val port = ((high shl BrowserDomainConfigureLimits.PORT_BYTE_SHIFT) or low).toUShort()
            val result = HelperResult.decode(payload.copyOf(BrowserDomainConfigureLimits.RESULT_SIZE_BYTES))
            return BrowserDomainConfigureResponse(result, port)
        }
    }
}

internal fun isCanonicalExactDomain(domain: String): Boolean {
    val length = domain.length
    if (length !in BrowserDomainConfigureLimits.MINIMUM_DOMAIN_LENGTH..BrowserDomainConfigureLimits.MAXIMUM_DOMAIN_LENGTH) {
        return false
    }
    if (!domain.all(::isCanonicalDomainCharacter)) {
        return false
    }
    val labels = domain.split('.', limit = Int.MAX_VALUE)
    val last = labels.lastOrNull() ?: return false
    return labels.size >= 2 &&
        last.any { character -> !character.isDigit() } &&
        labels.all(::isCanonicalDomainLabel)
}

private fun isCanonicalDomainCharacter(character: Char): Boolean {
    if (character.code >= BrowserDomainConfigureLimits.MAXIMUM_ASCII_CODE) {
        return false
    }
    return isDomainBodyCharacter(character)
}

private fun isDomainBodyCharacter(character: Char): Boolean {
    return character in 'a'..'z' || character.isDigit() || isDomainSymbol(character)
}

private fun isDomainSymbol(character: Char): Boolean {
    return character == '-' || character == '.'
}

private fun isCanonicalDomainLabel(label: String): Boolean {
    val length = label.length
    if (length !in BrowserDomainConfigureLimits.MINIMUM_LABEL_LENGTH..BrowserDomainConfigureLimits.MAXIMUM_LABEL_LENGTH) {
        return false
    }
    val first = label.first()
    val last = label.last()
    val validEdges = first.isLetterOrDigit() && last.isLetterOrDigit()
    return validEdges && label.all { character -> character.isLetterOrDigit() || character == '-' }
}
